### PR TITLE
[RISCV] Add CFI information for vector callee-saved registers

### PR DESCRIPTION
**Pull request by 4vtomat**
Currently the CFI offset for RVV registers are not handled entirely,
this patch add those information for either stack unwinding or
debugger to work correctly on RVV callee-saved stack object.

Depends On D154576

Differential Revision: https://reviews.llvm.org/D156846

### CODE REVIEW

**Comment by 4vtomat** [comment]
This is originally in this PR and I just split the two commits into two PRs, the first PR is merged, this is the second one.

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp [hunk 1]
```diff
@@ -452,29 +478,39 @@ static MCCFIInstruction createDefCFAExpression(const TargetRegisterInfo &TRI,
...
 
-  uint8_t buffer[16];
-  if (FixedOffset) {
-    Expr.push_back(dwarf::DW_OP_consts);
-    Expr.append(buffer, buffer + encodeSLEB128(FixedOffset, buffer));
-    Expr.push_back((uint8_t)dwarf::DW_OP_plus);
-    Comment << " + " << FixedOffset;
-  }
+  appendScalableVectorExpression(Expr, FixedOffset, ScalableOffset,
+                                 TRI.getDwarfRegNum(RISCV::VLENB, true),
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:482]: Why is the caller responsible for calling `TRI.getDwarfRegNum(RISCV::VLENB, true)`? Can we sink that into the function?

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp [hunk 2]
```diff
@@ -1532,6 +1583,63 @@ bool RISCVFrameLowering::spillCalleeSavedRegisters(
...
+  RISCVMachineFunctionInfo *RVFI = MF->getInfo<RISCVMachineFunctionInfo>();
+  const TargetInstrInfo &TII = *STI.getInstrInfo();
+  DebugLoc DL = MBB.findDebugLoc(MI);
+
+  const auto &RVVCSI = getRVVCalleeSavedInfo(*MF, MFI.getCalleeSavedInfo());
+  if (RVVCSI.empty())
+    return;
+
+  uint64_t FixedSize = getStackSizeWithRVVPadding(*MF) +
+                       RVFI->getLibCallStackSize() + RVFI->getRVPushStackSize();
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:1599]: I think getLibCallStackSize and getRVPushStackSize part of getStackSizeWithRVVPadding after 0de2b26942f890a6ec84cd75ac7abe3f6f2b2e37 last month

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp (hunk 1 above)
Comment (4vtomat) [reply llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:482]: Sure~

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp [hunk 3]
```diff
@@ -1532,6 +1582,62 @@ bool RISCVFrameLowering::spillCalleeSavedRegisters(
...
+          createDefCFAOffset(*STI.getRegisterInfo(), CS.getReg(), -FixedSize,
+                             MFI.getObjectOffset(FI) / 8));
+      BuildMI(MBB, MI, DL, TII.get(TargetOpcode::CFI_INSTRUCTION))
+          .addCFIIndex(CFIIndex)
+          .setMIFlag(MachineInstr::FrameSetup);
+    }
+  }
+}
+
+void RISCVFrameLowering::emitCalleeSavedRVVEpilogCFI(
```
Comment (kito-cheng) [inline llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:1620]: Don't emit CFI for epilogue yet, there is long history, but seriously I forgot the full story, my vaguely memory tell me it's some thing related to .cfi_remember_state and .cfi_restore_state so we don't emit any CFI for epilogue.

and I guess @asb and @jrtc27 may remember the story.

**Comment by 4vtomat** [comment]
Remove the epilog CFI generation.

File: llvm/test/CodeGen/RISCV/rvv/ctlz-vp.ll [hunk 4]
```diff
@@ -2615,6 +2615,7 @@ define <vscale x 16 x i64> @vp_ctlz_zero_undef_nxv16i64(<vscale x 16 x i64> %va,
 }
 
 define <vscale x 16 x i64> @vp_ctlz_zero_undef_nxv16i64_unmasked(<vscale x 16 x i64> %va, i32 zeroext %evl) {
+;
```
Comment (kito-cheng) [inline llvm/test/CodeGen/RISCV/rvv/ctlz-vp.ll:2618]: this blank line could remove?

**Review by kito-cheng — APPROVED** [review]
LGTM

**Pull request merged**